### PR TITLE
imp: build reproducibly

### PR DESCRIPTION
**Pull request by blaggacao**

### CODE REVIEW

**Comment by blaggacao** [comment]
<details>

```
> @cardano-sdk/wallet@0.3.0 build:cjs /nix/store/53nqbd53pwwhy3gw85arpfniw43mjgqd-__at__cardano-sdk__slash__wallet-0.3.0/lib/node_modules/@cardano-sdk/wall>
> tsc --build src

node_modules/@cardano-sdk/core/src/Asset/util/assetId.ts:23:46 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'AssetName' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"AssetName">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toStringTag]
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'AssetName' is not assignable to parameter of type 'string'.

23     assetName: CSL.AssetName.new(Buffer.from(assetName, 'hex')),
                                                ~~~~~~~~~


node_modules/@cardano-sdk/core/src/Asset/util/assetId.ts:24:55 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'PolicyId' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"PolicyId">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toStringTag]
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'PolicyId' is not assignable to parameter of type 'string'.

24     scriptHash: CSL.ScriptHash.from_bytes(Buffer.from(policyId, 'hex'))
                                                         ~~~~~~~~


node_modules/@cardano-sdk/core/src/Asset/util/metadatumToCip25.ts:71:65 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'AssetName' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'AssetName' is not assignable to parameter of type 'string'.

71     policy.get(asset.name.toString()) || policy.get(Buffer.from(asset.name, 'hex').toString('utf8'))
                                                                   ~~~~~~~~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/certificate.ts:99:39 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'VrfVkHex' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"VrfVkHex">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toStringTag]
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'VrfVkHex' is not assignable to parameter of type 'string'.

99     VRFKeyHash.from_bytes(Buffer.from(vrf, 'hex')),
                                         ~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/certificate.ts:107:93 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'Hash32ByteBase16<"Hash32ByteBase16">' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"Hash32ByteBase16">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toSt>
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'Hash32ByteBase16<"Hash32ByteBase16">' is not assignable to parameter of type 'string'.

107       ? PoolMetadata.new(URL.new(metadataJson.url), PoolMetadataHash.from_bytes(Buffer.from(metadataJson.hash, 'hex')))
                                                                                                ~~~~~~~~~~~~~~~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/coreToCsl.ts:85:63 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'TransactionId' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"TransactionId">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toStrin>
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'TransactionId' is not assignable to parameter of type 'string'.

85   TransactionInput.new(TransactionHash.from_bytes(Buffer.from(core.txId, 'hex')), core.index);
                                                                 ~~~~~~~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/coreToCsl.ts:156:54 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'PolicyId' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'PolicyId' is not assignable to parameter of type 'string'.

156       scriptHash = ScriptHash.from_bytes(Buffer.from(policyId, 'hex'));
                                                         ~~~~~~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/coreToCsl.ts:164:49 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'AssetName' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'AssetName' is not assignable to parameter of type 'string'.

164     mintAssets.insert(AssetName.new(Buffer.from(assetName, 'hex')), intQuantity);
                                                    ~~~~~~~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/coreToCsl.ts:194:60 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'Ed25519KeyHash' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"Ed25519KeyHash">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toStri>
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'Ed25519KeyHash' is not assignable to parameter of type 'string'.

194     cslKeyHashes.add(Ed25519KeyHash.from_bytes(Buffer.from(signature, 'hex')));
                                                               ~~~~~~~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/coreToCsl.ts:250:72 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'Hash32ByteBase16<"Hash32ByteBase16">' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'Hash32ByteBase16<"Hash32ByteBase16">' is not assignable to parameter of type 'string'.

250     cslBody.set_script_data_hash(ScriptDataHash.from_bytes(Buffer.from(scriptIntegrityHash, 'hex')));
                                                                           ~~~~~~~~~~~~~~~~~~~


node_modules/@cardano-sdk/core/src/CSL/coreToCsl/coreToCsl.ts:273:56 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'Ed25519PublicKey' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"Ed25519PublicKey">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toSt>
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'Ed25519PublicKey' is not assignable to parameter of type 'string'.

273     const publicKey = PublicKey.from_bytes(Buffer.from(vkey, 'hex'));
                                                           ~~~~


src/KeyManagement/cip8/cip30signData.ts:91:67 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'Ed25519PublicKey' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"Ed25519PublicKey">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toSt>
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'Ed25519PublicKey' is not assignable to parameter of type 'string'.

91   coseKey.set_header(CoseLabel.x, CBORValue.new_bytes(Buffer.from(publicKey, 'hex')));
                                                                     ~~~~~~~~~


src/KeyManagement/cip8/cip30signData.ts:111:17 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'HexBlob' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"HexBlob">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toStringTag]
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'HexBlob' is not assignable to parameter of type 'string'.

111     Buffer.from(payload, 'hex'),
                    ~~~~~~~


src/KeyManagement/cip8/cip30signData.ts:116:47 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'Ed25519Signature' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"Ed25519Signature">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toSt>
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'Ed25519Signature' is not assignable to parameter of type 'string'.

116   const coseSign1 = builder.build(Buffer.from(signature, 'hex'));
                                                  ~~~~~~~~~


src/KeyManagement/KeyAgentBase.ts:87:47 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'Bip32PublicKey' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
      Type 'OpaqueString<"Bip32PublicKey">' is missing the following properties from type 'SharedArrayBuffer': byteLength, [Symbol.species], [Symbol.toStri>
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'Bip32PublicKey' is not assignable to parameter of type 'string'.

87     const accountPublicKeyBytes = Buffer.from(this.extendedAccountPublicKey, 'hex');
                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


src/KeyManagement/InMemoryKeyAgent.ts:56:76 - error TS2769: No overload matches this call.
  Overload 1 of 4, '(arrayBuffer: ArrayBuffer | SharedArrayBuffer, byteOffset?: number | undefined, length?: number | undefined): Buffer', gave the followi>
    Argument of type 'HexBlob' is not assignable to parameter of type 'ArrayBuffer | SharedArrayBuffer'.
  Overload 2 of 4, '(str: string, encoding?: string | undefined): Buffer', gave the following error.
    Argument of type 'HexBlob' is not assignable to parameter of type 'string'.

56     const signature = Cardano.Ed25519Signature(signingKey.sign(Buffer.from(blob, 'hex')).to_hex());
                                                                              ~~~~
Found 16 errors.
```

</details>

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+          builtins.mapAttrs (n: _: let
+            otherNames = lib.remove n subpackagesNames;
+            otherSubpackages = builtins.removeAttrs subpackages [n];
+          in {
+            # install-method.installMethod = "copy";
+            add-inputs.buildInputs = old: old ++ (lib.attrValues topDependencies) ++ [pkgs.nodejs-14_x.pkgs.yarn];
+            install-symlinks.postInstall = "ln -s $out/lib/node_modules/${n}/dist $out/dist";
+            add-type-notation.prePatch = lib.optionalString (n == "@cardano-sdk/core")
+            ''
+              substituteInPlace ./src/Cardano/util/subtractValueQuantities.ts \
```
Comment (mkazlauskas) [inline flake.nix:43]: That's odd, why is this needed? We have many other relative imports from own package

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                "@cardano-sdk/rabbitmq" = [ "core" "util" ];
+                "@cardano-sdk/util" = [];
+                "@cardano-sdk/util-dev" = [ "core" ];
+                "@cardano-sdk/util-rxjs" = [];
+                "@cardano-sdk/wallet" = [ "cip2" "cip30" "core" "util-rxjs" "util"];
+                "@cardano-sdk/web-extension" = [ "cip30" "core" "wallet" "util-rxjs" "util"];
+              };
+              replacer = lib.concatMapStrings (name: ''
+                --replace \
+                  '"path": "../../${name}/src"' \
```
Comment (mkazlauskas) [inline flake.nix:66]: This should only be needed for `packages/[package]/test` projects. Are we running `tsc` for test projects? If we do I suggest omitting it to make maintenance easier

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                  '"path": "../node_modules/@cardano-sdk/${name}/src"' \
+              '');
+            in ''
+              cp ${./tsconfig.json} ./src/tsconfig-root.json
+              mkdir build
+              cp ${./build/cjs-package.json} ./build/cjs-package.json
```
Comment (mkazlauskas) [inline flake.nix:72]: Can we change package build script to from `build` to `build:cjs`? `build` is making 2 builds (into different module formats) and we only need 1 here.

Also, `cjs-package.json` and `esm-package.json` are not going to be needed then so we could simplify the nix file.

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                  '"extends": "../../../tsconfig.json"' \
+                  '"extends": "./tsconfig-root.json"' \
+              ${replacer references.${n}}
+
+            '';
+          })
+          subpackages
+        )
+        {
+          "chromedriver" = {
```
Comment (mkazlauskas) [inline flake.nix:90]: Are we running web extension e2e tests? :thinking:

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            dont-download = {
+                buildInputs = old: old ++ [pkgs.chromedriver];
+                CHROMEDRIVER_SKIP_DOWNLOAD = "true";
+            };
+          };
+          # "@cardano-sdk/cardano-services" = {
```
Comment (mkazlauskas) [inline flake.nix:96]: Is this going to be used?

File: packages/web-extension/src/keyAgent/consumeKeyAgent.ts [hunk 6]
```diff
@@ -1,3 +1,6 @@
+import '@cardano-sdk/util';
+import '@cardano-sdk/wallet/dist/cjs/KeyManagement';
+import 'rxjs';
```
Comment (mkazlauskas) [inline packages/web-extension/src/keyAgent/consumeKeyAgent.ts:3]: We need to do something about these imports, because they will be impacting end-users. One of the runtime environments of SDK is browser via bundlers. Bundle size matters and having such imports will prevent bundlers from doing tree-shaking (just include the entire libraries instead of selectively picking only the exports that are used)

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
Can we make it build only `cardano-services` and it's dependencies?

**Comment by blaggacao** [comment]
> Can we make it build only cardano-services and it's dependencies?

Counter-intuitively, scoping is in fact, harded to maintain than non-scoping due to the quality of the input data (that we can obtain from the `./locks/...` file).

File: flake.nix (hunk 1 above)
Comment (blaggacao) [reply flake.nix:43]: I have no good explanation. The TS compiler fails with the relative import claiming that it cannot infer xyz type. Without knowing for certain, it might be related to:

https://github.com/microsoft/TypeScript/issues/42873

(which does not contain qualified information)

File: flake.nix (hunk 2 above)
Comment (blaggacao) [reply flake.nix:66]: The relative location between packages is not guaranteed, and hence `../../other` can point to nowhere.

This is compartmentalization is actually an important feature for proper package-wise build caching.

File: flake.nix (hunk 3 above)
Comment (blaggacao) [reply flake.nix:72]: If these aren't going to be needed, I'd prefer them to be removed from the `build` entrypoint in `packages.json`. Packages json is our upstream here and any semantic deviation probably would require more (to be maintained) explanatory prose than we would gain by excluding this case.

If we remove this, anybody will immediately ask "why is this semantically different for no obvious reason"?

The second build never has made any problems, after all.

File: flake.nix (hunk 4 above)
Comment (blaggacao) [reply flake.nix:90]: This is not running the test, yet. But we _do_ build it, indeed. :small_airplane:

File: flake.nix (hunk 5 above)
Comment (blaggacao) [reply flake.nix:96]: Maybe (probably) yes, I should remove this for now, though... I haven't yet advanced to the entrypoint packaging, but I'm likely going to need it there.

File: flake.nix (hunk 1 above)
Comment (mkazlauskas) [reply flake.nix:43]: Seems related to needing to add those imports :thinking:

Comment (blaggacao) [reply flake.nix:43]: Yes, indeed. Maybe we can figure this out and eradicate the root cause. :pray:

File: flake.nix (hunk 2 above)
Comment (mkazlauskas) [reply flake.nix:66]: There shouldn't be any `../../other` imports outside of test code. Test code is only included in `[package]/test/tsconfig.json` and we should only be concerned about building with `[package]/src/tsconfig.json` unless we also want to build the tests (which is not even supported, we just use `ts-node` to run them)

File: flake.nix (hunk 3 above)
Comment (mkazlauskas) [reply flake.nix:72]: In context of SDK, `build` script prepares packages to be published to `npm` and we do want to support multiple module formats. Of course we could have a separate script for that, but it boils down to what is the primary purpose for these packages

**Comment by blaggacao** [comment]
closing in favor of a different take basend on yarn3 & `yarn-plugin-nixify`
